File: examples/09-ipc/polyglot/java/src/main/java/io/orionmesh/demo/JsSub.java
package io.orionmesh.demo;

import io.nats.client.*;
import io.nats.client.api.ConsumerConfiguration;
import io.nats.client.api.StreamConfiguration;
import io.nats.client.impl.NatsMessage;

import java.time.Duration;
import java.util.List;
import java.util.Map;

/**
 * Java JetStream subscriber — interoperates with the Rust + Python JS demos.
 *
 * Args:
 *   --nats-url nats://127.0.0.1:4222
 *   --subject  orion.demo.js
 *   --stream   ORION_DEMO_JS
 *   --durable  workers
 *
 * Pulls in batches of 10 with a 5-second wait; multiple subs sharing the same
 * durable name share the load and survive restart (replay from last ack).
 */
public class JsSub {
    public static void main(String[] argv) throws Exception {
        Map<String, String> args = parseArgs(argv);
        String url = args.getOrDefault("--nats-url",
                System.getenv().getOrDefault("NATS_URL", "nats://127.0.0.1:4222"));
        String subject = args.getOrDefault("--subject", "orion.demo.js");
        String streamName = args.getOrDefault("--stream", "ORION_DEMO_JS");
        String durable = args.getOrDefault("--durable", "workers");
        String label = args.getOrDefault("--label", labelFromEnv("java-js"));

        System.out.printf("[java-sub:%s] connecting to %s -> %s (JetStream stream=%s durable=%s)%n",
                label, url, subject, streamName, durable);
        try (Connection nc = Nats.connect(url)) {
            JetStreamManagement jsm = nc.jetStreamManagement();
            JetStream js = nc.jetStream();
            System.out.printf("[java-sub:%s] connected%n", label);

            String subjWildcard = (subject.contains("*") || subject.contains(">"))
                    ? subject : subject + ".>";
            try {
                jsm.addStream(StreamConfiguration.builder()
                        .name(streamName)
                        .subjects(subjWildcard)
                        .build());
            } catch (JetStreamApiException e) {
                // already exists — fine
            }
            System.out.printf("[java-sub:%s] stream %s ready%n", label, streamName);

            PullSubscribeOptions opts = PullSubscribeOptions.builder().durable(durable).build();
            JetStreamSubscription sub = js.subscribe(subjWildcard, opts);
            System.out.printf("[java-sub:%s] consumer '%s' bound%n", label, durable);

            while (true) {
                List<Message> msgs = sub.fetch(10, Duration.ofSeconds(5));
                for (Message m : msgs) {
                    String body = new String(m.getData());
                    long seq = m.metaData() != null ? m.metaData().streamSequence() : 0;
                    System.out.printf("[java-sub:%s] recv (seq=%d): %s (subject=%s)%n",
                            label, seq, body, m.getSubject());
                    m.ack();
                }
            }
        }
    }

    private static String labelFromEnv(String fallback) {
        String idx = System.getenv("ORION_REPLICA_INDEX");
        return idx != null ? "r" + idx : fallback;
    }

    private static Map<String, String> parseArgs(String[] argv) {
        var out = new java.util.HashMap<String, String>();
        for (int i = 0; i + 1 < argv.length; i += 2) out.put(argv[i], argv[i + 1]);
        return out;
    }
}

File: examples/09-ipc/polyglot/python/js_sub.py
#!/usr/bin/env python3
"""Python JetStream subscriber — interoperates with the Rust + Java JS demos.

Auto-creates the stream + a durable pull consumer; replays from last ack on
restart. Multiple subs sharing the same --durable name share the load.

Setup:
  bash setup.sh   # creates .venv, installs nats-py
"""
from __future__ import annotations

import argparse
import asyncio
import os
import sys

try:
    import nats
    from nats.js.api import StreamConfig, ConsumerConfig
except ImportError:
    sys.stderr.write("nats-py not installed. Run: pip install nats-py\n")
    sys.exit(1)


async def main():
    p = argparse.ArgumentParser()
    p.add_argument("--nats-url", default=os.environ.get("NATS_URL", "nats://127.0.0.1:4222"))
    p.add_argument("--subject", default="orion.demo.js")
    p.add_argument("--stream", default="ORION_DEMO_JS")
    p.add_argument("--durable", default="workers")
    p.add_argument("--label", default=None)
    args = p.parse_args()

    label = args.label or (
        f"r{os.environ.get('ORION_REPLICA_INDEX')}"
        if os.environ.get("ORION_REPLICA_INDEX")
        else "py-js"
    )

    print(
        f"[py-sub:{label}] connecting to {args.nats_url} -> {args.subject} "
        f"(JetStream stream={args.stream} durable={args.durable})",
        flush=True,
    )
    nc = await nats.connect(args.nats_url)
    js = nc.jetstream()
    print(f"[py-sub:{label}] connected", flush=True)

    subj_wildcard = args.subject if ("*" in args.subject or ">" in args.subject) else f"{args.subject}.>"
    try:
        await js.add_stream(StreamConfig(name=args.stream, subjects=[subj_wildcard]))
    except Exception:
        pass
    print(f"[py-sub:{label}] stream {args.stream} ready", flush=True)

    # Pull subscription using a durable consumer name.
    psub = await js.pull_subscribe(
        f"{args.subject}.>",
        durable=args.durable,
        stream=args.stream,
    )
    print(f"[py-sub:{label}] consumer '{args.durable}' bound", flush=True)

    while True:
        try:
            msgs = await psub.fetch(batch=10, timeout=5)
        except Exception:
            # timeout — keep polling
            continue
        for m in msgs:
            body = m.data.decode("utf-8", errors="replace")
            seq = m.metadata.sequence.stream if m.metadata else 0
            print(f"[py-sub:{label}] recv (seq={seq}): {body} (subject={m.subject})", flush=True)
            await m.ack()


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        pass
